cpp/util/etcd.cc

```diff
@@ -803,101 +737,81 @@ EtcdClient::~EtcdClient() {
 }
 
 
-bool EtcdClient::MaybeUpdateLeader(const libevent::HttpRequest& req,
-                                   Request* etcd_req) {
-  // We're talking to the leader, get back to normal processing...
-  if (req.GetResponseCode() != 307) {
-    return false;
-  }
-
-  const char* const location(
-      CHECK_NOTNULL(evhttp_find_header(req.GetInputHeaders(), "location")));
+void EtcdClient::FetchDone(Request* etcd_req, util::Task* task) {
+  VLOG(2) << "EtcdClient::FetchDone: " << task->status();
 
-  const unique_ptr<evhttp_uri, void (*)(evhttp_uri*)> uri(
-      evhttp_uri_parse(location), &evhttp_uri_free);
-  CHECK(uri);
-  LOG(INFO) << "etcd leader: " << evhttp_uri_get_host(uri.get()) << ":"
-            << evhttp_uri_get_port(uri.get());
-
-  // Update the last known leader, and retry the request on the new
-  // leader.
-  etcd_req->Run(UpdateLeader(evhttp_uri_get_host(uri.get()),
-                             evhttp_uri_get_port(uri.get())));
+  if (!task->status().ok()) {
+    // TODO(pphaneuf): If there is a connection problem, we should
```

Comment:
I did the same thing, but it's a bit far in the diff, we'll see how that goes...